src/lib/io/sol-uart-impl-riot.c

```diff
@@ -29,57 +29,58 @@
 #include "sol-mainloop.h"
 #include "sol-util-internal.h"
 #include "sol-interrupt_scheduler_riot.h"
+#include "sol-util.h"
+#include "sol-vector.h"
 
 SOL_LOG_INTERNAL_DECLARE_STATIC(_log_domain, "uart");
 
 struct sol_uart {
-    uart_t id;
-    struct {
-        void *handler;
-
-        void (*rx_cb)(void *data, struct sol_uart *uart, uint8_t byte_read);
-        const void *rx_user_data;
-
-        void (*tx_cb)(void *data, struct sol_uart *uart, uint8_t *tx, int status);
-        const void *tx_user_data;
-        struct sol_timeout *tx_writer;
-        const uint8_t *tx_buffer;
-        unsigned int tx_length;
-    } async;
+    void (*rx_cb)(void *data, struct sol_uart *uart);
+    void (*tx_cb)(void *data, struct sol_uart *uart, struct sol_blob *blob, int status);
+    const void *user_data;
+    struct sol_timeout *tx_writer;
+    void *handler;
+    struct sol_ptr_vector pending_blobs;
+    struct sol_buffer rx_buffer;
+    int id;
 };
 
 static void
 uart_rx_cb(void *arg, uint8_t data)
 {
     struct sol_uart *uart = arg;
+    int r;
 
-    if (!uart->async.rx_cb)
-        return;
-    uart->async.rx_cb((void *)uart->async.rx_user_data, uart,
-        (uint8_t)data);
+    r = sol_buffer_append_char(&uart->rx_buffer, (char)data);
+    SOL_INT_CHECK(r, < 0);
+    uart->rx_cb((void *)uart->user_data, uart);
```

Comment:
what if we use a 0-timeout to dispatch this? Then we can accumulate all that comes from `uart_rx_cb` (scheduled from uart thread) and dispatch to user in one go. It can be something like "if not 0-timeout scheduled, schedule one".